Neos.Flow/Classes/Mvc/Routing/Dto/RouteTags.php

```diff
@@ -0,0 +1,145 @@
+<?php
+namespace Neos\Flow\Mvc\Routing\Dto;
+
+/*
+ * This file is part of the Neos.Flow package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use Neos\Flow\Annotations as Flow;
+
+/**
+ * RouteTags to be associated with matched/resolved routes
+ *
+ * RouteTags can be added by Route Part handlers via the ResolveResult/MatchResult return values
+ * The tags will be added to the corresponding cache entries, so that they can be flushed selectively using the RouterCachingService
+ *
+ * @Flow\Proxy(false)
+ */
+final class RouteTags
+{
+
+    /**
+     * Pattern a tag must match. @see \Neos\Cache\Frontend\FrontendInterface::PATTERN_TAG
+     */
+    const PATTERN_TAG = '/^[a-zA-Z0-9_%\-&]{1,250}$/';
+
+    /**
+     * @var string[] numeric array of strings satisfying the PATTERN_TAG regex
+     */
+    private $tags = [];
+
+    /**
+     * @param string[] $tags numeric array of strings satisfying the PATTERN_TAG regex
+     */
+    private function __construct(array $tags)
+    {
+        $this->tags = $tags;
+    }
+
+    /**
+     * Creates an instance without any tags
+     *
+     * @return RouteTags
+     */
+    public static function createEmpty(): self
+    {
+        return new static([]);
+    }
+
+    /**
+     * Creates an instance with one given tag
+     *
+     * @param string $tag Tag value satisfying the PATTERN_TAG regex
+     * @return RouteTags
+     */
+    public static function createFromTag(string $tag): self
+    {
+        self::validateTag($tag);
+        return new static([$tag]);
+    }
+
+    /**
+     * Creates an instance with one given tags
+     *
+     * @param string[] $tags An array of strings satisfying the PATTERN_TAG regex
+     * @return RouteTags
+     */
+    public static function createFromArray(array $tags): self
+    {
+        array_walk($tags, 'static::validateTag');
+        return new static($tags);
+    }
+
+    /**
+     * Merges two instances of this class combining and unifying all tags
+     *
+     * @param RouteTags $tags
+     * @return RouteTags
+     */
+    public function merge(RouteTags $tags): self
+    {
+        $mergedTags = array_unique(array_merge($this->tags, $tags->tags));
+        return new static($mergedTags);
+    }
+
+    /**
+     * Creates a new instance with the given $tag added
+     * If the $tag has been added already, this instance is returned
+     *
+     * @param string $tag
+     * @return RouteTags
+     */
+    public function withTag(string $tag): self
```

Comment:
RouteTags::add can be more in sync with method like RouteTags::has

Reply:
Well… It returns `self`, for `add()` I'd expect a `bool` or `void` return value. It's in line with e.g. how PSR-7 handles mutators.